### PR TITLE
Add list-rules and list-worksheet-rules to conditionalformat (#730)

### LINKED ISSUE #730: Add list-rules action to conditionalformat for reading existing rules
## Summary

Add a `list-rules` (or `get-rules`) action to the `conditionalformat` tool that allows reading existing conditional formatting rules from a range or worksheet. Currently only `add-rule` and `clear-rules` are supported — there is no way to inspect what rules already exist.

## Motivation

- **Round-trip safety**: Before modifying or clearing rules, an LLM or script needs to check what is already applied
- **Debugging**: Users want to verify that rules were set correctly
- **Migration**: Copying rules from one range/sheet to another requires reading them first
- **Audit**: Understanding why certain cells are formatted a certain way

### Option A: `list-rules` action (per range)

```json
{
  "action": "list-rules",
  "sheetName": "Sheet1",
  "rangeAddress": "A1:G41"
}
```

**Response:**
```json
{
  "success": true,
  "rules": [
    {
      "type": "cellValue",
      "operator": "greater",
      "formula1": "100",
      "appliesTo": "$A$1:$A$41",
      "interiorColor": "#FFFF00",
      "fontColor": null,
      "fontBold": null
    },
    {
      "type": "expression",
      "formula1": "=$G1>1000",
      "appliesTo": "$A$1:$G$41",
      "interiorColor": "#FF0000",
      "fontColor": "#FFFFFF",
      "fontBold": true
    }
  ]
}
```

### Option B: `list-worksheet-rules` action (entire sheet)

```json
{
  "action": "list-worksheet-rules",
  "sheetName": "Sheet1"
}
```

Returns all rules across all ranges on the sheet, grouped by range.

## COM API Reference

The Excel COM `FormatConditions` collection (accessible via `Range.FormatConditions`) supports enumeration with properties:
- `Type` — rule type enum (xlCellValue, xlExpression, etc.)
- `Operator` — operator enum (xlBetween, xlEqual, xlGreater, etc.)
- `Formula1` / `Formula2` — formula strings
- `AppliesTo` — range the rule applies to (accessible via `FormatCondition.AppliesTo`)
- `Interior.Color` / `Font.Color` / `Font.Bold` / `Font.Italic` / `Borders` — formatting properties

## Considerations

- Color values: return as `#RRGGBB` hex strings for consistency with existing `add-rule` input format
- Priority: rules are ordered by priority in the collection — preserve this order in the response
- StopIfTrue: optionally expose the `StopIfTrue` property
- Performance: for large worksheets with many rules, consider pagination or limiting to a specific range

## Implementation Hints

- New action handled in `ConditionalFormattingCommands.cs`
- New result model `ConditionalFormatListResult` in `src/ExcelMcp.Core/Models/`
- CLI command: `conditionalformat list-rules --session <id> --sheet-name "Sheet1" --range-address "A1:Z100"`
- Reuse existing `ParseConditionalFormattingType` / `ParseConditionalFormattingOperator` for the reverse mapping (int → string)

## Acceptance Criteria

- [ ] `list-rules` action added to `conditionalformat` tool (MCP + CLI parity)
- [ ] Returns rule type, operator, formulas, applies-to range, and formatting (interior/font/borders)
- [ ] Supports both per-range and per-worksheet scoping
- [ ] Color values returned as `#RRGGBB` hex strings
- [ ] Rules returned in priority order
- [ ] CLI: `conditionalformat list-rules` and `conditionalformat list-worksheet-rules` commands added
- [ ] Tests for success and edge cases (no rules, invalid range, empty sheet)

### PR DESCRIPTION
**Pull request by sbroenne**
## Summary

Closes #730. Adds two read actions to the `conditionalformat` tool so existing conditional formatting rules can be inspected — previously only `add-rule` and `clear-rules` were available.

- **`list-rules`** — reads rules for a range (`Range.FormatConditions`)
- **`list-worksheet-rules`** — reads all rules on a sheet (`Worksheet.Cells.FormatConditions`)

Both return each rule''s type, operator, formulas, applies-to range, priority, `stopIfTrue`, and formatting (interior/font/borders), with colors as `#RRGGBB` hex strings, in priority order. MCP + CLI parity is generated automatically from `IConditionalFormattingCommands`.

## Changes

- **New models**: `ConditionalFormatListResult` + `ConditionalFormatRuleInfo` (all formatting fields nullable — only populated when the rule sets them).
- **Helpers**: `FormattingHelpers.ColorToHex` (reverse of `ParseColor`), plus reverse int→string mappers for rule type, operator, and border style.
- **Commands**: `ListRules` / `ListWorksheetRules` on the interface + implementations, using the existing `dynamic?` + `finally`/`ComUtilities.Release` COM pattern. Color-scale/data-bar/icon-set rules degrade gracefully (type only, null formatting).
- **Docs**: FEATURES.md (Conditional Formatting 2 → 4 ops; total 232 → 234), README/plugin/gh-pages docs, and `skills/shared/conditionalformat.md`.
- **Changeset**: `.changeset/conditionalformat-list-rules.md` (minor).

## Tests

New `ConditionalFormattingCommandsTests` — 8 real-Excel integration tests, all passing: no rules (empty range + empty sheet), single cellValue rule, expression rule with font+interior, multiple rules in priority order, worksheet-scope aggregation, invalid sheet/range error paths.

## Acceptance criteria

- [x] `list-rules` action added (MCP + CLI parity)
- [x] Returns rule type, operator, formulas, applies-to range, and formatting
- [x] Supports both per-range and per-worksheet scoping
- [x] Colors returned as `#RRGGBB` hex strings
- [x] Rules returned in priority order
- [x] CLI: `list-rules` and `list-worksheet-rules` commands added
- [x] Tests for success and edge cases (no rules, invalid range, empty sheet)

## Notes

Excel normalizes numeric `cell-value` formulas, so `100` reads back as `=100`. The raw Excel value is returned intentionally — stripping the leading `=` would corrupt expression rules like `=$G1>1000`.

All 14 pre-commit gates pass locally.

### CODE REVIEW

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 1]
```diff
@@ -157,8 +157,263 @@ public OperationResult ClearRules(
...
+
+                // Interior (fill)
+                try
+                {
+                    interior = fc.Interior;
+                    int colorIndex = interior.ColorIndex;
+                    if (colorIndex != -4142) // xlColorIndexNone
+                    {
+                        rule.InteriorColor = FormattingHelpers.ColorToHex((int)interior.Color);
+                        try { rule.InteriorPattern = (int)interior.Pattern; }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs]: Good catch. Fixed in 78de4cd: these numeric COM props are now read via Convert.ToInt32(..., InvariantCulture) so values surfaced as double no longer throw. I also added InvalidCastException to the COM read guard, since unset props can come back as DBNull (which Convert.ToInt32 rejects) - that case is now skipped cleanly.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 2]
```diff
@@ -157,8 +157,263 @@ public OperationResult ClearRules(
...
+                catch (Exception ex) when (IsComOrBinderException(ex)) { }
+
+                // Font
+                try
+                {
+                    font = fc.Font;
+                    int fontColorIndex = font.ColorIndex;
+                    if (fontColorIndex != -4105 && fontColorIndex != -4142) // not Automatic/None
+                    {
+                        try { rule.FontColor = FormattingHelpers.ColorToHex((int)font.Color); }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs]: Good catch. Fixed in 78de4cd: these numeric COM props are now read via Convert.ToInt32(..., InvariantCulture) so values surfaced as double no longer throw. I also added InvalidCastException to the COM read guard, since unset props can come back as DBNull (which Convert.ToInt32 rejects) - that case is now skipped cleanly.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 3]
```diff
@@ -157,8 +157,263 @@ public OperationResult ClearRules(
...
+                // Borders (read the left edge border as representative)
+                try
+                {
+                    borders = fc.Borders;
+                    edgeBorder = borders.Item(7); // xlEdgeLeft
+                    int lineStyle = edgeBorder.LineStyle;
+                    if (lineStyle != -4142) // xlLineStyleNone
+                    {
+                        rule.BorderStyle = BorderStyleToString(lineStyle) ?? lineStyle.ToString(System.Globalization.CultureInfo.InvariantCulture);
+                        try { rule.BorderColor = FormattingHelpers.ColorToHex((int)edgeBorder.Color); }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs]: Good catch. Fixed in 78de4cd: these numeric COM props are now read via Convert.ToInt32(..., InvariantCulture) so values surfaced as double no longer throw. I also added InvalidCastException to the COM read guard, since unset props can come back as DBNull (which Convert.ToInt32 rejects) - that case is now skipped cleanly.

File: tests/ExcelMcp.Core.Tests/Integration/Commands/ConditionalFormat/ConditionalFormattingCommandsTests.cs [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+
+        Assert.True(result.Success);
+        Assert.Equal(2, result.Rules.Count);
+        // Priority values, when present, should be ascending in collection order.
+        var priorities = result.Rules
+            .Where(r => r.Priority.HasValue)
+            .Select(r => r.Priority!.Value)
+            .ToList();
+        var sorted = priorities.OrderBy(p => p).ToList();
+        Assert.Equal(sorted, priorities);
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply tests/ExcelMcp.Core.Tests/Integration/Commands/ConditionalFormat/ConditionalFormattingCommandsTests.cs]: Fixed in 78de4cd. cellValue rules always carry a priority, so the test now asserts every rule has one (Assert.All ... Priority.HasValue) before checking ascending order, instead of filtering nulls out.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 5]
```diff
@@ -157,8 +157,264 @@ public OperationResult ClearRules(
...
+        }
+
+        return rules;
+    }
+
+    private static string ReadRuleType(dynamic fc)
+    {
+        try { return ConditionalFormattingTypeToString((int)fc.Type); }
+        catch (Exception ex) when (IsComOrBinderException(ex)) { return "unknown"; }
+    }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs:359]: Fixed in the latest commit - fc.Type and fc.Operator are now read via Convert.ToInt32(..., InvariantCulture), so a double-surfaced value no longer degrades the rule to unknown/null.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 6]
```diff
@@ -157,8 +157,264 @@ public OperationResult ClearRules(
...
+
+    private static string? ReadRuleOperator(dynamic fc)
+    {
+        try { return ConditionalFormattingOperatorToString((int)fc.Operator); }
+        catch (Exception ex) when (IsComOrBinderException(ex)) { return null; }
+    }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs:365]: Fixed in the latest commit - fc.Type and fc.Operator are now read via Convert.ToInt32(..., InvariantCulture), so a double-surfaced value no longer degrades the rule to unknown/null.

File: src/ExcelMcp.Core/Models/ConditionalFormatTypes.cs [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+    /// Conditional formatting rules in priority order.
+    /// </summary>
+    public List<ConditionalFormatRuleInfo> Rules { get; set; } = [];
+}
+
+/// <summary>
+/// Describes a single conditional formatting rule read from Excel.
+/// Formatting properties are only populated when the rule actually sets them.
+/// </summary>
+public class ConditionalFormatRuleInfo
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Models/ConditionalFormatTypes.cs:31]: Done - the file is now named ConditionalFormatTypes.cs, matching the ResultTypes.cs / PivotTableTypes.cs convention for multi-type model files.

File: src/ExcelMcp.Core/Models/ConditionalFormatListResult.cs [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+    /// Font italic state, when set (null = not specified / mixed).
+    /// </summary>
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+    public bool? FontItalic { get; set; }
+
+    /// <summary>
+    /// Border style name, when set (applied to edge borders).
+    /// </summary>
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+    public string? BorderStyle { get; set; }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Models/ConditionalFormatListResult.cs:112]: Updated: the XML doc now states BorderStyle/BorderColor are read by scanning the edge borders (left/top/bottom/right) and using the first one that has a style, matching the read implementation.

File: src/ExcelMcp.Core/Models/ConditionalFormatListResult.cs [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+
+    /// <summary>
+    /// Border color as #RRGGBB, when set.
+    /// </summary>
+    [JsonIgnore(Condition = JsonIgnoreCondition.WhenWritingNull)]
+    public string? BorderColor { get; set; }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Models/ConditionalFormatListResult.cs:119]: Updated: the XML doc now states BorderStyle/BorderColor are read by scanning the edge borders (left/top/bottom/right) and using the first one that has a style, matching the read implementation.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 10]
```diff
@@ -157,8 +157,264 @@ public OperationResult ClearRules(
...
+                    edgeBorder = borders.Item(7); // xlEdgeLeft
+                    int lineStyle = Convert.ToInt32(edgeBorder.LineStyle, System.Globalization.CultureInfo.InvariantCulture);
+                    if (lineStyle != -4142) // xlLineStyleNone
+                    {
+                        rule.BorderStyle = BorderStyleToString(lineStyle) ?? lineStyle.ToString(System.Globalization.CultureInfo.InvariantCulture);
+                        try { rule.BorderColor = FormattingHelpers.ColorToHex(Convert.ToInt32(edgeBorder.Color, System.Globalization.CultureInfo.InvariantCulture)); }
+                        catch (Exception ex) when (IsComOrBinderException(ex)) { }
+                    }
+                }
+                catch (Exception ex) when (IsComOrBinderException(ex)) { }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs]: Good point - fixed. ReadFormatConditions now scans all four edge borders (left/top/bottom/right) and uses the first non-none edge, so rules that set only top/bottom/right borders are reported correctly.

File: tests/ExcelMcp.Core.Tests/Integration/Commands/ConditionalFormat/ConditionalFormattingCommandsTests.cs [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+        var result = _commands.ListRules(batch, "", "A1:G41");
+
+        Assert.True(result.Success);
+        var rule = Assert.Single(result.Rules);
+        Assert.Equal("expression", rule.Type);
+        Assert.Equal("=$G1>1000", rule.Formula1);
+        Assert.Equal("#FF0000", rule.InteriorColor);
+        Assert.Equal("#FFFFFF", rule.FontColor);
+        Assert.True(rule.FontBold);
+    }
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply tests/ExcelMcp.Core.Tests/Integration/Commands/ConditionalFormat/ConditionalFormattingCommandsTests.cs:83]: I added a border round-trip test but discovered AddRule's conditional-format border write path throws a COM error (Unable to set the LineStyle property of the Border class) - a pre-existing limitation with no prior coverage, outside this read-focused PR's scope. So a clean round-trip test via the command surface isn't currently feasible. The read logic itself was hardened here (all four edges, Convert.ToInt32). I'll file a follow-up for the AddRule CF border write.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 12]
```diff
@@ -157,8 +157,272 @@ public OperationResult ClearRules(
...
+
+                // Interior (fill)
+                try
+                {
+                    interior = fc.Interior;
+                    int colorIndex = Convert.ToInt32(interior.ColorIndex, System.Globalization.CultureInfo.InvariantCulture);
+                    if (colorIndex != -4142) // xlColorIndexNone
+                    {
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs:293]: Fixed - Interior.ColorIndex == -4105 (xlColorIndexAutomatic) is now treated the same as -4142 (None), so InteriorColor is left null when no explicit fill was set, matching the font-color logic.

File: src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs [hunk 13]
```diff
@@ -157,8 +157,272 @@ public OperationResult ClearRules(
...
+    /// <summary>
+    /// Reads a FormatConditions collection into a list of rule descriptors.
+    /// Each optional COM property read is guarded so unsupported rule types degrade gracefully.
+    /// </summary>
+    private static List<ConditionalFormatRuleInfo> ReadFormatConditions(dynamic formatConditions)
+    {
+        var rules = new List<ConditionalFormatRuleInfo>();
+
+        int count = formatConditions.Count;
+        for (int i = 1; i <= count; i++)
```
[the remark this answers is not in the document]
Comment (sbroenne) [reply src/ExcelMcp.Core/Commands/ConditionalFormat/ConditionalFormattingCommands.cs:256]: Fixed - formatConditions.Count is now read via Convert.ToInt32(..., InvariantCulture), so a double-surfaced count no longer breaks rule enumeration. This completes the COM-numeric-read hardening across the CF read path (Count, Type, Operator, ColorIndex/Color, Pattern, LineStyle, Priority).
